Security Cipher Module: Enabled Backup and Restore support for NOS Upgrades

nmoray opened the pull request:
<!--
   Enabled Backup and Restore support in Security Cipher Module for NOS Upgrades
   Reference PR: https://github.com/sonic-net/sonic-buildimage/pull/17201
-->

#### Why I did it
This is required majorly for NOS upgrade scenario to backup cipher_pass file stored at /etc dir.

#### How I did it
Added pre/post migration scripts

#### How to verify it
NOS upgrade


#### Which release branch to backport (provide reason below if selected)
- [ ] 202205
- [ ] 202211
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505

#### Tested branch (Please provide the tested image version)
- [x] 202311

#### Link to config_db schema for YANG module changes
https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/doc/Configuration.md

nmoray commented:
@qiluo-msft please help me understand [ms_conflict](https://github.com/sonic-net/sonic-buildimage/pull/22711/checks?check_run_id=42970042984) failure. I don't know why this is failing. 
Additionally, please approve the PR in case there are no new comments.

anders-nexthop commented:
@qiluo-msft @lguohan are there any further comments on this PR? If not, can we merge it and unblock this feature?

@nmoray this is what the MS-Conflict says:
```
failure: @StormLiangMS,@liushilongbuaa Conflict already exists in master
Please wait a few hours to run ms_conflict again!
'/azpw ms_conflict'
```
I don't know what this check is for, but maybe just rebase and push again to re-run it and see if that resolves things?

mssonicbld commented:
/azp run Azure.sonic-buildimage

nmoray commented:
@anders-nexthop @qiluo-msft Please review the updated code and share the comments if any.

anders-nexthop commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -12,89 +12,140 @@
+            syslog.syslog(syslog.LOG_ERR, "rotate_feature_passwd: No callbacks registered for {}".format(feature_type))
+            return
+
+        if new_password is not None:
```
so if new_password is None, we will still rotate but using the existing password (so we will still trigger all the callbacks). Do I understand that correctly?

nmoray replied:
You are right. This is help in synchronising the passkeys for a given feature in case one of them got out of sync.

anders-nexthop reviewed (commented):
Looks good, the rotation code is a good change.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -12,89 +12,140 @@
+            with open(CIPHER_PASS_FILE, 'r') as f:
+                return json.load(f)
+        except Exception as e:
+            syslog.syslog(syslog.LOG_ERR, "del_cipher_passwd: Exception occurred: {}".format(e))
```
yeah

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

anders-nexthop commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -12,89 +12,141 @@
-    
-
-    def encrypt_passkey(self, feature_type, secret: str, passwd: str) -> str:
+    def rotate_feature_passwd(self, feature_type, table_info, secret, new_password=None):
```
why are we passing the secret here? Isn't the secret per-instance, while only the password is shared between users of a feature? I don't think we want to combine `rotate()` with any actual instance secret updates.

anders-nexthop commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -12,89 +12,141 @@
+
+        syslog.syslog(syslog.LOG_INFO, "rotate_feature_passwd: Rotating password for feature {} and notifying callbacks...".format(feature_type))
+        for cb in callbacks:
+            cb(table_info, secret)
```
I wonder if there is a more reliable way of handling this callback functionality. We expect the encrypted secrets to be stored in ConfigDB, correct? Would we ever want the callback to do anything other than update entries in specified table paths? maybe instead of actual callbacks, we can store table paths that need to be updated and in this function we can either iterate through them and update each one or maybe build a patch and use GCU apply-patch to update them all cleanly in one operation.

nmoray replied:
As per the discussion, now replaced callbacks with just table info so that whenever any feature registers, the respective table_info will be attached to the Feature_id and same will be used while rotating the password.

anders-nexthop commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -12,89 +12,141 @@
+            return
+
+        if new_password is not None:
+            data[feature_type]["password"] = new_password
```
I think we are missing something here. we expect the user callback to update the encrypted secrets stored in ConfigDB, but those secrets have been encrypted with the old password. we can't just encrypt them again with the new one, we have to first decrypt each secret using the old password, then re-encrypt it using the new one and store it back in ConfigDB. Otherwise we will lose the original secret and things won't work properly.

nmoray replied:
As per https://github.com/sonic-net/sonic-utilities/pull/3936 implementation, user is not allowed to rotate the password with specifying the passkey or secret. So we don't need to worry about decryption and encryption.

anders-nexthop commented on src/sonic-yang-models/yang-models/sonic-system-tacacs.yang:
```diff
@@ -91,6 +91,12 @@ module sonic-system-tacacs {
                     description "Authentication type";
                 }
 
+                leaf key_encrypt {
```
Would this be better as a typedef, so that multiple features can use the same type?

anders-nexthop replied:
The advantage to a template would be that we would have some measure of certainty that implementers would all be using the same format when adding the key_encrypt leaf. The disadvantage would be that every feature would have to use the same default, but I think given the implementation that would be okay.

This is subjective. I would have a slight preference for having a shared template as the base, but I'm not that familiar with SONiC's YANG best practices. Up to you if you think this is helpful to add or not.

nmoray commented on src/sonic-py-common/tests/test_security_cipher.py:
```diff
@@ -114,23 +115,6 @@ def test_encrypt_raises_if_no_password(self):
             with pytest.raises(ValueError):
                 temp.encrypt_passkey("RADIUS", "plaintext")
 
-    def test_is_key_encrypt_enabled(self):
```
Added the testcase back.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -146,39 +99,123 @@ def decrypt_passkey(self, feature_type,  b64_encoded: str) -> str:
+                data[feature_type]["table_info"].remove(table_info)
+                if not data[feature_type]["table_info"]:
+                    # No more table_info left; remove password as well
+                    data[feature_type]["password"] = None
```
There is no harm in removing the feature entry itself if no single registered entry exist. But we need to delete the cipher_pass file as well if there are no registered features too. Just to avoid that I kept the footprints of the registered Features. 
If we decide to have only the password without table_info then rotate feature won't be possible. IMO, now the new design with Rotate feature has been settled down. :-) 

Why do you think the coupling of password and table_info will cause an issue? If so, what will it be?

anders-nexthop replied:
To be clear, I think that we absolutely want the rotation mechanism, and I think the overall design is very close to correct at this point!

Thinking ahead to how this service might be used in the field, I can imagine having a startup script that configures the encryption password for a feature when the device is first provisioned, with some rotation schedule set after that initial setup. The feature itself may be enabled or disabled many times (or have the passkey added/removed many times) before that rotation schedule is up. I believe with the current design, we will remove the table_info for a feature when that feature is globally disabled, and each time we do so we will force the user to either re-enter the password or use a new one. Managing this may be challenging; if the user wants to use the same password as they had before (which they might want if they are using the same encryption password on several devices, so that they can use a common config with a single encrypted feature password for all devices), they now need to keep track of that password and re-apply it as needed.

My feeling is that we would rather have the encryption password be something that is not reset very frequently (only when you explicitly want to rotate), while the table_info can come and go as needed.

Here's how it could look (very similar to what you have already):

* register()/deregister(): adds/removes tables associated with a given feature. No change to feature password. table_info can be empty and password will remain set.
* add_feature_password()/rotate_feature_password(): sets/rotates feature password. On rotate, any passkeys from associated table_info(s) will be updated. If there are no associated tables, the feature password is still updated.
* del_feature_password(): if we need, we can add back the option to remove the feature password entirely. This would be like a rotation, where we would go and decrypt all the associated passkeys, but without the re-encryption step. This would disable the encryption service, but would preserve configured passkeys. The list of table_info would remain intact, so if the service is re-enabled at some point the configured passkeys will be updated accordingly. I don't know if we actually need or want this method; if the user has a passkey with key_encrypt set but the service is disabled, they will have an unencrypted passkey even though they have requested an encrypted one, which doesn't seem good. I include a description of how this would work mainly just to demonstrate the mechanics of how we could maintain the table_info and the feature passkey separately.

To me, this is slightly better as it makes the contract with the caller simpler. You can update the feature password, or you can update the list of tables associated with a feature, or you can encrypt/decrypt. Each operation is separate. We can let the caller decide if they want to rotate the key when the feature is enabled/disabled. I think that this encryption service may end up being pretty generally useful in sonic, so having fewer internal assumptions is beneficial.

Does this suggest make sense or is it still not clear what I'm getting at? Do you see any downsides of the suggested approach that I am not considering?

nmoray replied:
As per the current design, the user has to enter the password only when he/ she wants to encrypt the passkey for the first time and second time when he / she wants to rotate or change the feature password.  If the user wants to use the same password, he / she can keep that password and keep encrypting the new passkey as many times as he / she wants. 

Please refer to the following unit test logs.
```
1. Encrypt passkey:
sonic# config tacacs passkey nikhil --encrypt
sonic# show run al | jq .TACPLUS
{
  "global": {
    "auth_type": "login",
    "key_encrypt": "True",
    "passkey": "U2FsdGVkX19YTUcG+0r3+d78A5E0zTXEukS3ZNrfyFk=",
    "src_intf": "Loopback0"
  }
}
sonic# cat /etc/pam.d/common-auth-sonic
#THIS IS AN AUTO-GENERATED FILE
#
# /etc/pam.d/common-auth- authentication settings common to all services
# This file is included from other service-specific PAM config files,
# and should contain a list of the authentication modules that define
# the central authentication scheme for use on the system
# (e.g., /etc/shadow, LDAP, Kerberos, etc.). The default is to use the
# traditional Unix authentication mechanisms.
#
# here are the per-package modules (the "Primary" block)

auth	[success=done new_authtok_reqd=done default=ignore auth_err=die]	pam_tacplus.so server=<> secret=nikhil login=login timeout=5   try_first_pass
auth	[success=done new_authtok_reqd=done default=ignore auth_err=die]	pam_tacplus.so server=<> secret=nikhil login=login timeout=5   try_first_pass
auth	[success=1 default=ignore]	pam_unix.so nullok try_first_pass

#
# here's the fallback if no module succeeds
auth    requisite                       pam_deny.so
# prime the stack with a positive return value if there isn't one already;
# this avoids us returning an error just because nothing sets a success code
# since the modules above will each just jump around
auth    required                        pam_permit.so
# and here are more per-package modules (the "Additional" block)
sonic# cat /etc/cipher_pass.json
{
  "TACPLUS": {
    "table_info": [
      "TACPLUS|global"
    ],
    "password": "TEST1"
  }

2. Encrypt passkey and update existing password
sonic# config tacacs passkey nikhil --encrypt --rotate
Password:
sonic# cat /etc/cipher_pass.json
{
  "TACPLUS": {
    "table_info": [
      "TACPLUS|global"
    ],
    "password": "TEST3"
  }
sonic# cat /etc/pam.d/common-auth-sonic
#THIS IS AN AUTO-GENERATED FILE
#
# /etc/pam.d/common-auth- authentication settings common to all services
# This file is included from other service-specific PAM config files,
# and should contain a list of the authentication modules that define
# the central authentication scheme for use on the system
# (e.g., /etc/shadow, LDAP, Kerberos, etc.). The default is to use the
# traditional Unix authentication mechanisms.
#
# here are the per-package modules (the "Primary" block)

auth	[success=done new_authtok_reqd=done default=ignore auth_err=die]	pam_tacplus.so server=<> secret=nikhil login=login timeout=5   try_first_pass
auth	[success=done new_authtok_reqd=done default=ignore auth_err=die]	pam_tacplus.so server=<> secret=nikhil login=login timeout=5   try_first_pass
auth	[success=1 default=ignore]	pam_unix.so nullok try_first_pass

#
# here's the fallback if no module succeeds
auth    requisite                       pam_deny.so
# prime the stack with a positive return value if there isn't one already;
# this avoids us returning an error just because nothing sets a success code
# since the modules above will each just jump around
auth    required                        pam_permit.so
# and here are more per-package modules (the "Additional" block)
sonic# show run al | jq .TACPLUS
{
  "global": {
    "auth_type": "login",
    "key_encrypt": "True",
    "passkey": "U2FsdGVkX18/8nFrAiCe01pOqgPoUUZFaTpmtawdi8I=",
    "src_intf": "Loopback0"
  }
}

3. Revert back to plaintext passkey
sonic# config tacacs passkey ravi
sonic# cat /etc/pam.d/common-auth-sonic
#THIS IS AN AUTO-GENERATED FILE
#
# /etc/pam.d/common-auth- authentication settings common to all services
# This file is included from other service-specific PAM config files,
# and should contain a list of the authentication modules that define
# the central authentication scheme for use on the system
# (e.g., /etc/shadow, LDAP, Kerberos, etc.). The default is to use the
# traditional Unix authentication mechanisms.
#
# here are the per-package modules (the "Primary" block)

auth	[success=done new_authtok_reqd=done default=ignore auth_err=die]	pam_tacplus.so server=<> secret=ravi login=login timeout=5   try_first_pass
auth	[success=done new_authtok_reqd=done default=ignore auth_err=die]	pam_tacplus.so server=<> secret=ravi login=login timeout=5   try_first_pass
auth	[success=1 default=ignore]	pam_unix.so nullok try_first_pass

#
# here's the fallback if no module succeeds
auth    requisite                       pam_deny.so
# prime the stack with a positive return value if there isn't one already;
# this avoids us returning an error just because nothing sets a success code
# since the modules above will each just jump around
auth    required                        pam_permit.so
# and here are more per-package modules (the "Additional" block)
sonic# show run al | jq .TACPLUS
{
  "global": {
    "auth_type": "login",
    "key_encrypt": "False",
    "passkey": "ravi",
    "src_intf": "Loopback0"
  }
}
sonic# cat /etc/cipher_pass.json
{
  "TACPLUS": {
    "table_info": [],
    "password": null
  }
```

nmoray replied:
<When the user reverts back to using a plaintext password, they lose the configured encryption password>. It is expected, isn't it! :-) If the user reverts back to plaintext passkey. There should not be any password for the respective feature. Later if the user wants to encrypt that passkey, he / she must provide the passkey along with a password. 
If we keep the old password and used that for the new encryption, then it won't right as we will be using stale data in that case.

In case of a reboot, the configs gets removed (unsaved ones) but the password saved in cipher_pass.json file won't be removed unless and until the default config set the key_encrypt flag to false.

anders-nexthop replied:
The reboot case you mentioned points out the issue I'm describing perfectly. I can configure a passkey on a running system and pass in an encryption key. If I don't save the running config and reboot the system, it will start up again without the passkey, but the encryption key will still be saved. We are then in the case that we are trying to avoid, where there is a stale encryption key being used to encrypt passkeys. So no matter what, the encryption key handling is somewhat disconnected from the passkey handling, and we rely on the caller knowing about this case and handling it correctly. I think we would be better served to make that difference explicit by providing a `delete_feature_password()` that can be called explicitly by the user to disable the encryption for that feature. This way, the contract provided to the caller is very clear; they must explicitly `set` and `delete` the feature password manually when they want to enable/disable the service for a feature, and can use `encrypt` and `decrypt` only on an enabled feature.

This seems to me to be a cleaner approach, with less corner-case ambiguity. However I am happy to advocate for this change as a later enhancement, should the benefit become more clear as we implement callers. Let's keep pushing the feature forward and enhance it as needed, rather than try to solve for every eventuality here.

anders-nexthop replied:
Right, but if it's understood that the feature password is not managed through ConfigDB then that seems more predictable from a user standpoint. I set the encryption passkey for a feature on the device, and I have not unset it, therefore I am not surprised when it still exists.

Since we added the migration logic, which preserves configured passwords even across upgrades, it seems like we generally want to keep the encryption passwords around regardless of current configuration. Otherwise why bother with the migration at all?

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -146,39 +99,123 @@ def decrypt_passkey(self, feature_type,  b64_encoded: str) -> str:
+                # Update DB
+                db_entry["passkey"] = new_encrypted_passkey
+                # Make sure key_encrypt should be set true
+                db_entry["key_encrypt"] = 'True'
```
Please refer to https://github.com/sonic-net/sonic-utilities/pull/3936/files. I have handled this from the Feature side and added this explicitly to avoid any failure due to a corner case. From the feature side, I have made sure that this method has to be called if Encrypt is enabled. So I am explicitly adding key_encrypt flag to avoid a case where there is no key_encrypt flag for a registered TABLE.

anders-nexthop replied:
I don't think we want to push logic into the caller if we can avoid it, since that makes this service less general. At the moment, we are just thinking about AAA users, but there's no reason we couldn't use this mechanism to store any type of sensitive identifying information in ConfigDB. We'd be better off to solve corner cases in the API code if we can, so that we don't have to re-implement the same logic for every caller.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -146,39 +99,123 @@ def decrypt_passkey(self, feature_type,  b64_encoded: str) -> str:
+    def encrypt_passkey(self, feature_type, secret: str) -> str:
         """
-        Removes only the password for the given feature_type while keeping the file structure intact.
+        Encrypts the plaintext and returns the result as a hex string.
```
As it's a wrapper, I have not added the Algorithm related details in this. _encrypt_passkey() is the actual API where everything happens so I have added those details there.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

anders-nexthop commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -164,24 +164,23 @@ def rotate_feature_passwd(self, feature_type, new_password):
             encrypted_passkey = db_entry.get("passkey")
-            if encrypted_passkey:
+            #Rotate only if valid passkey is present and 'key_encyrpt' flag is True
+            if encrypted_passkey and db_entry.get("key_encrypt") == 'True':
```
I always struggle with the yang boolean type. Isn't the key_encrypt string supposed to 'true'?

nmoray replied:
Agreed!

anders-nexthop reviewed (approved):
I am approving this change, as I think this is a workable implementation and would like to see it move forward. However, please consider my point about separating encryption password handling from table registration.

@qiluo-msft can you take another look, or assign another reviewer to get more eyes on this latest version?

@bhouse-nexthop your thoughts would be welcome here given your security background.

mssonicbld commented:
/azp run Azure.sonic-buildimage

nmoray commented:
Thanks @anders-nexthop for the diligent review.
@qiluo-msft @bhouse-nexthop I have addressed the comments. Please approve in case there are no more comments.

nmoray commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

nmoray commented:
@qiluo-msft will you please help in merging this PR in case there are no more comments?

anders-nexthop commented:
@qiluo-msft and @lguohan can we get this PR merged if there are no more comments?

lguohan commented:
do we have sonic-mgmt test to test this?

qiluo-msft commented on files/image_config/config-setup/01-post-security-cipher:
```diff
@@ -0,0 +1,11 @@
+# Restore cipher_pass file from persistent storage
+if [ -f /host/security_cipher/cipher_pass ]; then
+    cp /host/security_cipher/cipher_pass.json /etc/cipher_pass.json
+    chmod 640 /etc/cipher_pass.json
```
>640 [](http://example.com/codeflow?start=10&length=3)

Is it better to use `600` if the only reader is root?

nmoray replied:
sure.

nmoray commented:
> do we have sonic-mgmt test to test this?

@lguohan No, but once this PR is merged, I will create a new PR to cover the sonic-mgmt testcases.

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -28,69 +31,33 @@ def __new__(cls):
+        try:
+            with open(CIPHER_PASS_FILE, 'r') as f:
+                return json.load(f)
+        except Exception as e:
```
@qiluo-msft Like? Any specific exception that you are looking for?

Is "except json.JSONDecodeError" okay? or "except (json.JSONDecodeError, OSError)"? 

IMO, "Exception as e" will catch all the runtime exceptions and also print the respective error in the form of 'e'. What do you think?

mssonicbld commented:
/azp run Azure.sonic-buildimage

nmoray commented:
@qiluo-msft comments addressed. Please check and approve if no more comments.

nmoray commented on files/image_config/config-setup/01-pre-security-cipher:
```diff
@@ -0,0 +1,13 @@
+mkdir -p /host/security_cipher
+
+# Copy cipher_pass file to persistent storage
+if [ -f /etc/cipher_pass ]; then
```
done

nmoray commented on src/sonic-py-common/sonic_py_common/security_cipher.py:
```diff
@@ -146,39 +134,122 @@ def decrypt_passkey(self, feature_type,  b64_encoded: str) -> str:
+
+    def set_feature_password(self, feature_type, password):
+        """
+        Set a new password for a feature type.
```
done

mssonicbld commented:
/azp run Azure.sonic-buildimage
